ambry-network/src/main/java/com.github.ambry.network/NetworkClient.java

```diff
@@ -85,17 +92,27 @@ public NetworkClient(Selector selector, NetworkConfig networkConfig, int maxConn
    */
   public List<ResponseInfo> sendAndPoll(List<RequestInfo> requestInfos)
       throws IOException {
-    if (closed) {
-      throw new IllegalStateException("The NetworkClient is closed.");
-    }
-    List<ResponseInfo> responseInfoList = new ArrayList<ResponseInfo>();
-    for (RequestInfo requestInfo : requestInfos) {
-      pendingRequests.add(new RequestMetadata(time.milliseconds(), requestInfo));
+    long startTime = time.milliseconds();
+    try {
+      if (closed) {
+        throw new IllegalStateException("The NetworkClient is closed.");
+      }
+      List<ResponseInfo> responseInfoList = new ArrayList<ResponseInfo>();
+      numPendingConnections.set(pendingRequests.size());
+      for (RequestInfo requestInfo : requestInfos) {
+        ClientNetworkRequestMetrics clientNetworkRequestMetrics =
+            new ClientNetworkRequestMetrics(networkMetrics.requestQueueTime, networkMetrics.requestSendTime,
+                networkMetrics.requestSendTotalTime, 0);
+        pendingRequests.add(new RequestMetadata(time.milliseconds(), requestInfo, clientNetworkRequestMetrics));
+        numPendingConnections.set(pendingRequests.size());
+      }
+      List<NetworkSend> sends = prepareSends(responseInfoList);
+      selector.poll(POLL_TIMEOUT_MS, sends);
+      handleSelectorEvents(responseInfoList);
+      return responseInfoList;
+    } finally {
+      networkMetrics.networkClientSendAndPollTime.update(time.milliseconds() - startTime);
```

Comment:
You shouldn't at line 107, and that is my point. In order for the metric to make sense, these requests should not be considered as pending unless they remain in the pending list after the prepare.